sdk/projects/azure-projects/azure/projects/_resource.py

```diff
@@ -0,0 +1,632 @@
+    azd_dir = os.path.join(os.getcwd(), ".azure")
+    if name and not name.endswith(".env"):
+        if not os.path.isdir(azd_dir):
+            return {}
+        env_name = name + (f"-{label}" if label else "")
+        env_path = os.path.join(azd_dir, env_name, ".env")
+    elif name:
+        env_path = name
+    else:
+        script_name = os.path.splitext(os.path.basename(inspect.stack()[-1][0].f_code.co_filename))[0]
+        script_env = os.path.join(azd_dir, script_name, ".env")
+        if os.path.isdir(azd_dir) and os.path.isfile(script_env):
+            env_path = script_env
+        else:
+            env_path = ".env"
+    values = dotenv_values(env_path)
+    return values
+
+
+def _build_envs(services: List[str], attributes: List[str]) -> List[str]:
+    all_vars = product(services, attributes)
+    return ["_".join(["AZURE"] + list(var)).upper() for var in all_vars]
+
+
+@runtime_checkable
+class SyncClient(Protocol, ContextManager["SyncClient"]):
+    def close(self) -> None: ...
+
+
+@runtime_checkable
+class AsyncClient(Protocol, AsyncContextManager["AsyncClient"]):  # pylint: disable=async-client-bad-name
+    async def close(self) -> None: ...
+
+
+class FieldType(NamedTuple):
+    resource: str
+    identifier: ResourceIdentifiers
+    version: str
+    properties: Dict[str, Any]
+    symbol: ResourceSymbol
+    outputs: Dict[str, List[Output]]
+    resource_group: Optional[ResourceSymbol]
+    extensions: "ExtensionResources"
+    existing: bool
+    name: Optional[Union[str, Parameter]]
+    defaults: Optional[Dict[str, Any]]
+
+
+class ResourceReference(TypedDict, total=False):
+    name: Union[str, Parameter]
+    resource_group: "ResourceGroup[ResourceReference]"
+    subscription: Union[str, Parameter]
+    parent: Resource
+
+
+class ExtensionResources(TypedDict, total=False):
+    managed_identity_roles: Union[
+        Parameter,
+        Sequence[Union[Parameter, "RoleAssignment", str]],
+    ]
+    user_roles: Union[
+        Parameter,
+        Sequence[Union[Parameter, "RoleAssignment", str]],
+    ]
+    # lock
+    # diagnostics
+    # private endpoint
+    # secret store?
+
+
+FieldsType = Dict[str, FieldType]
+ClientType = TypeVar("ClientType", bound=Union[SyncClient, AsyncClient])
+_EMPTY_DEFAULT: MutableMapping[str, Any] = {}
+_EMPTY_DEFAULT_EXTENSIONS: ExtensionResources = {}
+
+
+class Resource:  # pylint: disable=too-many-instance-attributes
+    DEFAULTS: MutableMapping[str, Any] = _EMPTY_DEFAULT
+    DEFAULT_EXTENSIONS: ExtensionResources = _EMPTY_DEFAULT_EXTENSIONS
+    identifier: ResourceIdentifiers
+    properties: Mapping[str, Any]
+    extensions: ExtensionResources
+    parent: Optional[Resource]
+
+    def __init__(self, properties: Optional[Mapping[str, Any]] = None, /, **kwargs) -> None:
+        """This constructor should not be used directly."""
+        self.properties = properties or {}
+        self.extensions: ExtensionResources = kwargs.pop("extensions", {})
```

Comment:
Why not in the method signature?

Reply:
This was mostly to "hide" things from the public API docs, because this object is public, but people shouldn't use it directly - so I didn't want the complex constructor showing up in the docs.
Though I don't feel too strongly about it. Maybe I could even just make this base class internal...